File: library-beta/src/main/java/com/useek/library_beta/USeekPlayerView.java
package com.useek.library_beta;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.os.Build;
import android.util.AttributeSet;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.FrameLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import java.net.URL;

/**
 * This is custom layout which you can easily drop in XML resource or create anywhere in your code.
 *
 * There are 2 ways to use USeekPlayerView.
 *
 * 1. Add to XML resource layout file.
 *
 *        com.useek.library_beta.USeekPlayerView
 *        android:id="@+id/useek_player_view"
 *        android:layout_width="match_parent"
 *        android:layout_height="match_parent"
 *
 * 2. Add as a subview programmatically
 *
 *       useekView = new USeekPlayerView(this);
 *       useekContainer.addView(
 *           useekView.getView(),
 *               new LinearLayout.LayoutParams(
 *                   ViewGroup.LayoutParams.MATCH_PARENT,
 *                   ViewGroup.LayoutParams.MATCH_PARENT
 *               )
 *           );
 *       useekView.loadVideo("{mGameId}", "{mUserId"});
 *       useekView.setPlayerListener(this);
 *
 */
public class USeekPlayerView extends FrameLayout {

    private Context mContext;
    private final String mTagName = getResources().getString(R.string.library_name);

    private WebView mWebView;
    private FrameLayout mLoadingMaskView;
    private TextView mLoadingTextView;
    private View mCustomView;
    private LayoutInflater mInflater;

    private String mGameId;
    private String mUserId;
    private String mLoadingText;

    private VideoLoadStatus mStatus = VideoLoadStatus.NONE;
    private Boolean mIsLoadingMaskHidden = false;

    private USeekPlayerListener mPlayerListener;


    public USeekPlayerView(Context context) {
        super(context);
        mInflater = (LayoutInflater)context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        init(context, null, 0);
    }

    public USeekPlayerView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs, 0);
    }

    public USeekPlayerView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(context, attrs, defStyle);
    }

    private void init(Context context, AttributeSet attrs, int defStyle) {

        mContext = context;

        View view = LayoutInflater.from(context).inflate(R.layout.useek_view, this, true);
        mWebView = view.findViewById(R.id.useek_web_view);
        mWebView.getSettings().setLoadWithOverviewMode(true);
        mWebView.getSettings().setUseWideViewPort(true);

        mLoadingMaskView = view.findViewById(R.id.useek_loading_mask_view);
        mLoadingTextView = view.findViewById(R.id.useek_loading_text);
        ProgressBar progressBar = view.findViewById(R.id.useek_view_progressbar);
        progressBar.getIndeterminateDrawable().setColorFilter(0xAAA, android.graphics.PorterDuff.Mode.SRC_ATOP);
        this.mIsLoadingMaskHidden = false;

        // Load attributes
        final TypedArray a = getContext().obtainStyledAttributes(
                attrs, R.styleable.USeekPlayerView, defStyle, 0);

        if (a.hasValue(R.styleable.USeekPlayerView_useek_loadingText)) {
            this.mLoadingText = a.getString(R.styleable.USeekPlayerView_useek_loadingText);
            this.mLoadingTextView.setText(this.mLoadingText);
        }

        a.recycle();
    }

    /**
     * Set placeholder text for loading video
     * @param loadingText     String to set placeholder text while loading video
     */
    public void setLoadingText(String loadingText) {
        this.mLoadingText = loadingText;
    }

    /**
     * Create USeekPlayerView as subview programmatically
     *
     * @return instance object of USeekPlayerView
     */
    public View getView() {
        if (mCustomView == null) {
            mCustomView = mInflater.inflate(R.layout.useek_view, null);
            mWebView = mCustomView.findViewById(R.id.useek_web_view);
            mWebView.getSettings().setLoadWithOverviewMode(true);
            mWebView.getSettings().setUseWideViewPort(true);

            mLoadingMaskView = mCustomView.findViewById(R.id.useek_loading_mask_view);
            mLoadingTextView = mCustomView.findViewById(R.id.useek_loading_text);
            ProgressBar progressBar = mCustomView.findViewById(R.id.useek_view_progressbar);
            progressBar.getIndeterminateDrawable().setColorFilter(0xAAA, android.graphics.PorterDuff.Mode.SRC_ATOP);

            this.mIsLoadingMaskHidden = false;
        }
        return mCustomView;
    }

    /**
     * Set unique game id provided by USeek
     * @param gameId    String object of unique game id provided by USeek, Non-nullable
     */
    public void setGameId(String gameId) {
        this.mGameId = gameId;
    }

    /**
     * Get unique game id
     * @return mGameId   String object of unique game id
     */
    public String getGameId() {
        return mGameId;
    }

    /**
     * Set user's unique id registered on USeek
     * @param userId    String object of user's unique id registered on USeek, Nullable
     */
    public void setUserId(String userId) {
        this.mUserId = userId;
    }

    /**
     * Get user's unique id registered on USeek
     * @return mUserId   String object of user's unique id
     */
    public String getUserId() {
        return mUserId;
    }

    /**
     * Set listener for video loading mStatus
     * @param playerListener    USeekPlayerListener instance
     */
    public void setPlayerListener(USeekPlayerListener playerListener) {
        this.mPlayerListener = playerListener;
    }

    private USeekPlayerListener getPlayerListener() {
        return mPlayerListener;
    }

    /**
     * Load Video with current game id and user id
     */
    public void loadVideo() {
        this.loadVideo(this.mGameId, this.mUserId);
    }

    /**
     * Load Video with game id and user id
     * @param gameId    unique game id provided by USeek, not nullable
     * @param userId    user's unique id registered in USeek, nullable
     */
    public void loadVideo(String gameId, String userId) {
        this.setGameId(gameId);
        this.setUserId(userId);

        if (!this.validateConfiguration()) return;

        URL url = this.generateVideoUrl();
        if (url == null) {
            Log.e(mTagName, "Invalid USeek URL");
            return;
        }

        /** Initialize WebView */
        this.initializeWebView();

        /** Set mWebView listener */
        final USeekPlayerView _this = this;
        mWebView.setWebViewClient(new WebViewClient() {
            @Override
            public void onPageStarted(WebView view, String url, Bitmap favicon) {
                super.onPageStarted(view, url, favicon);
                _this.didStartLoadingWebView();
            }

            @Override
            public void onPageFinished(WebView view, String url) {
                super.onPageFinished(view, url);
                _this.didFinishLoadingWebView();
            }

            @Override
            public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
                super.onReceivedError(view, request, error);
                _this.didFailLoadingWebView(error);
            }
        });

        /** Load Video URL */
        mWebView.loadUrl(url.toString());

    }

    /**
     * Generate Video url
     * @return url object with generated video url with publisher id, game id and user id
     */
    private URL generateVideoUrl() {
        if (this.validateConfiguration()) {
            String urlString = String.format("https://www.useek.com/sdk/1.0/%s/%s/play",
                    USeekManager.sharedInstance().getPublisherId(),
                    this.getGameId());
            if (this.getUserId() != null && this.getUserId().length() > 0) {
                urlString = urlString + String.format("?external_user_id=%s", this.getUserId());
            }
            try {
                URL url = new URL(urlString);
                return url;
            } catch (Exception e) {
                return null;
            }
        } else {
            return null;
        }
    }

    /**
     * Check validation for publish id and game id
     * @return boolean for validate information
     */
    private Boolean validateConfiguration() {
        Boolean isValid = true;

        String publisherId = USeekManager.sharedInstance().getPublisherId();

        if (publisherId == null) {
            isValid = false;
            Log.e(mTagName, "Not set publisher id");
        } else if (publisherId.length() == 0){
            isValid = false;
            Log.e(mTagName, "Invalid publisher id");
        }

        if (this.getGameId() == null) {
            isValid = false;
            Log.e(mTagName, "Not set game id");
        } else if (this.getGameId().length() == 0) {
            isValid = false;
            Log.e(mTagName, "Invalid game id");
        }
        return isValid;
    }

    private void initializeWebView() {
        if (mWebView != null) {
            WebSettings webSettings = mWebView.getSettings();
            webSettings.setJavaScriptEnabled(true);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                webSettings.setMediaPlaybackRequiresUserGesture(false);
            }
            mWebView.setBackgroundColor(Color.TRANSPARENT);

        } else {
            Log.e(mTagName, "Not initialized web view.");
        }
    }

    /**
     * Started loading video(web view)
     */
    private void didStartLoadingWebView() {
        Log.d(mTagName, "WebView didStartLoadingWebView");

        if (this.mStatus == VideoLoadStatus.NONE) {
            if (this.mPlayerListener != null) {
                this.mPlayerListener.useekPlayerDidStartLoad(this);
            }
        }

        this.mStatus = VideoLoadStatus.LOAD_STARTED;
        if (!this.mIsLoadingMaskHidden) {
            this.animateLoadingMaskToShow();
        } else {
            this.mLoadingMaskView.setVisibility(GONE);
        }

    }

    /**
     * Finished loading video(web view)
     */
    private void didFinishLoadingWebView() {
        Log.d(mTagName, "WebView didFinishLoadingWebView");

        if (this.mStatus != VideoLoadStatus.LOAD_FAILED &&
                this.mStatus != VideoLoadStatus.LOADED) {
            if (this.mPlayerListener != null) {
                this.mPlayerListener.useekPlayerDidFinishLoad(this);
            }
        }

        if (this.mStatus != VideoLoadStatus.LOAD_FAILED) {
            this.mStatus = VideoLoadStatus.LOADED;
        }

        if (!this.mIsLoadingMaskHidden) {
            this.animateLoadingMaskToHide();
        } else {
            this.mLoadingMaskView.setVisibility(GONE);
        }

    }

    /**
     * Failed loading video(web view)
     */
    private void didFailLoadingWebView(WebResourceError error) {
        String errorString = error.toString();
        if (errorString == null) {
            errorString = "";
        }
        Log.d(mTagName, "WebView didFailLoadingWebView with Error:\n" + errorString);

        if (this.mStatus != VideoLoadStatus.LOAD_FAILED) {
            if (this.mPlayerListener != null) {
                this.mPlayerListener.useekPlayerDidFailWithError(this, error);
            }
        }

        this.mStatus = VideoLoadStatus.LOAD_FAILED;

        if (!this.mIsLoadingMaskHidden) {
            this.animateLoadingMaskToHide();
        } else {
            this.mLoadingMaskView.setVisibility(GONE);
        }
    }

    /**
     * Animation of showing mLoadingMaskView
     */
    private void animateLoadingMaskToShow() {
        this.mLoadingMaskView.setVisibility(VISIBLE);
    }

    /**
     * Animation of hiding mLoadingMaskView
     */
    private void animateLoadingMaskToHide() {
        this.mLoadingMaskView.setVisibility(GONE);
    }


    enum VideoLoadStatus {
        NONE,           // Not stated loading
        LOAD_STARTED,   // Started loading
        LOADED,         // Completed loading
        LOAD_FAILED,    // Failed loading
    }
}

File: library-kt-beta/src/main/java/com/useek/library_kt_beta/USeekPlayerView.kt
package com.useek.library_kt_beta

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Color
import android.os.Build
import android.util.AttributeSet
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.FrameLayout
import android.widget.ProgressBar
import android.widget.TextView
import java.net.URL

/**
 * This is custom layout which you can easily drop in XML resource or create anywhere in your code.
 *
 * There are 2 ways to use USeekPlayerView.
 *
 * 1. Add to XML resource layout file.
 *
 *        com.useek.library_kt_beta.USeekPlayerView
 *        android:id="@+id/playerView"
 *        android:layout_width="match_parent"
 *        android:layout_height="match_parent"
 *
 * 2. Add as a subview programmatically
 *
 *       useekView = new USeekPlayerView(this);
 *       useekContainer.addView(
 *           useekView.getView(),
 *               new LinearLayout.LayoutParams(
 *                   ViewGroup.LayoutParams.MATCH_PARENT,
 *                   ViewGroup.LayoutParams.MATCH_PARENT
 *               )
 *           );
 *       useekView.loadVideo("{mGameId}", "{mUserId"});
 *       useekView.setPlayerListener(this);
 *
 */
class USeekPlayerView : FrameLayout {

    var gameId: String? = null
    var userId: String? = null
    var loadingText: String? = null

    var status = VideoLoadStatus.NONE
    var isLoadingMaskHidden: Boolean = false

    var playerListener: USeekPlayerListener? = null

    private var mContext: Context? = null
    private val mTagName = resources.getString(R.string.library_name)

    private var mWebView: WebView? = null
    private lateinit var mLoadingMaskView: FrameLayout
    private lateinit var mLoadingTextView: TextView
    private var mCustomView: View? = null
    private lateinit var mInflater: LayoutInflater

    constructor(context: Context) : super(context) {
        init(context, null, 0)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(context, attrs, 0)
    }

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        init(context, attrs, defStyle)
    }

    private fun init(context: Context, attrs: AttributeSet?, defStyle: Int) {
        mContext = context

        val view = LayoutInflater.from(context).inflate(R.layout.useek_view, this, true)
        mWebView = view.findViewById(R.id.useek_web_view)
        mWebView!!.settings.loadWithOverviewMode = true
        mWebView!!.settings.useWideViewPort = true

        mLoadingMaskView = view.findViewById(R.id.loadingMaskView)
        mLoadingTextView = view.findViewById(R.id.loadingText)
        val progressBar = view.findViewById<ProgressBar>(R.id.progressBar)
        progressBar.indeterminateDrawable.setColorFilter(0xAAA, android.graphics.PorterDuff.Mode.SRC_ATOP)
        this.isLoadingMaskHidden = false

        // Load attributes
        val a = getContext().obtainStyledAttributes(
                attrs, R.styleable.USeekPlayerView, defStyle, 0)

        if (a.hasValue(R.styleable.USeekPlayerView_useek_loadingText)) {
            this.loadingText = a.getString(R.styleable.USeekPlayerView_useek_loadingText)
            this.mLoadingTextView.text = this.loadingText
        }

        a.recycle()
    }

    /**
     * Create USeekPlayerView as subview programmatically
     *
     * @return instance object of USeekPlayerView
     */
    fun getView(): View {
        if (mCustomView == null) {
            mCustomView = mInflater.inflate(R.layout.useek_view, null)
            mWebView = mCustomView!!.findViewById(R.id.useek_web_view)
            mWebView!!.settings.loadWithOverviewMode = true
            mWebView!!.settings.useWideViewPort = true

            mLoadingMaskView = mCustomView!!.findViewById(R.id.useek_loading_mask_view)
            mLoadingTextView = mCustomView!!.findViewById(R.id.useek_loading_text)
            val progressBar = mCustomView!!.findViewById<ProgressBar>(R.id.useek_view_progressbar)
            progressBar.indeterminateDrawable.setColorFilter(0xAAA, android.graphics.PorterDuff.Mode.SRC_ATOP)

            this.isLoadingMaskHidden = false
        }
        return mCustomView!!
    }

    /**
     * Load Video with current game id and user id
     */
    fun loadVideo() {
        if (gameId != null && userId != null)
            this.loadVideo(this.gameId!!, this.userId!!)
    }

    /**
     * Load Video with game id and user id
     * @param gameId    unique game id provided by USeek, not nullable
     * @param userId    user's unique id registered in USeek, nullable
     */
    fun loadVideo(gameId: String, userId: String) {
        this.gameId = gameId
        this.userId = userId

        if ((!this.validateConfiguration())!!) return

        val url = this.generateVideoUrl()
        if (url == null) {
            Log.e(mTagName, "Invalid USeek URL")
            return
        }

        val webView = mWebView
        if (webView == null) {
            Log.e(mTagName, "Not created web view.")
            return
        }

        /** Initialize WebView  */
        this.initializeWebView()

        /** Set mWebView listener  */
        val me = this
        webView.webViewClient = object : WebViewClient() {
            override fun onPageStarted(view: WebView, url: String, favicon: Bitmap) {
                super.onPageStarted(view, url, favicon)
                me.didStartLoadingWebView()
            }

            override fun onPageFinished(view: WebView, url: String) {
                super.onPageFinished(view, url)
                me.didFinishLoadingWebView()
            }

            override fun onReceivedError(view: WebView, request: WebResourceRequest, error: WebResourceError) {
                super.onReceivedError(view, request, error)
                me.didFailLoadingWebView(error)
            }
        }

        /** Load Video URL  */
        webView.loadUrl(url.toString())
    }

    /**
     * Destroy view
     */
    fun destroy() {
        this.removeView(mWebView)
        mWebView?.destroy()
        mWebView = null
    }


    /**
     * Generate Video url
     * @return url object with generated video url with publisher id, game id and user id
     */
    private fun generateVideoUrl(): URL? {
        return if (this.validateConfiguration()) {
            var urlString = String.format("https://www.useek.com/sdk/1.0/%s/%s/play",
                    USeekManager.sharedInstance.publisherId,
                    this.gameId)
            if (this.userId != null && this.userId!!.isNotEmpty()) {
                urlString += String.format("?external_user_id=%s", this.userId)
            }
            try {
                URL(urlString)
            } catch (e: Exception) {
                null
            }
        } else {
            null
        }
    }

    /**
     * Check validation for publish id and game id
     * @return boolean for validate information
     */
    private fun validateConfiguration(): Boolean {
        var isValid = true

        val publisherId = USeekManager.sharedInstance.publisherId

        if (publisherId == null) {
            isValid = false
            Log.e(mTagName, "Not set publisher id")
        } else if (publisherId.isEmpty()) {
            isValid = false
            Log.e(mTagName, "Invalid publisher id")
        }

        if (this.gameId == null) {
            isValid = false
            Log.e(mTagName, "Not set game id")
        } else if (this.gameId!!.isEmpty()) {
            isValid = false
            Log.e(mTagName, "Invalid game id")
        }
        return isValid
    }

    @SuppressLint("SetJavaScriptEnabled")
    private fun initializeWebView() {
        val webView = mWebView
        if (webView != null) {
            val webSettings = webView.settings
            webSettings.javaScriptEnabled = true
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                webSettings.mediaPlaybackRequiresUserGesture = false
            }
            webView.setBackgroundColor(Color.TRANSPARENT)

        } else {
            Log.e(mTagName, "Not initialized web view.")
        }
    }

    /**
     * Started loading video(web view)
     */
    private fun didStartLoadingWebView() {
        Log.d(mTagName, "WebView didStartLoadingWebView")

        if (this.status == VideoLoadStatus.NONE) {
            if (this.playerListener != null) {
                this.playerListener!!.useekPlayerDidStartLoad(this)
            }
        }

        this.status = VideoLoadStatus.LOAD_STARTED
        if (!this.isLoadingMaskHidden) {
            this.animateLoadingMaskToShow()
        } else {
            this.mLoadingMaskView.visibility = View.GONE
        }

    }

    /**
     * Finished loading video(web view)
     */
    private fun didFinishLoadingWebView() {
        Log.d(mTagName, "WebView didFinishLoadingWebView")

        if (this.status != VideoLoadStatus.LOAD_FAILED && this.status != VideoLoadStatus.LOADED) {
            if (this.playerListener != null) {
                this.playerListener!!.useekPlayerDidFinishLoad(this)
            }
        }

        if (this.status != VideoLoadStatus.LOAD_FAILED) {
            this.status = VideoLoadStatus.LOADED
        }

        if (!this.isLoadingMaskHidden) {
            this.animateLoadingMaskToHide()
        } else {
            this.mLoadingMaskView.visibility = View.GONE
        }

    }

    /**
     * Failed loading video(web view)
     */
    private fun didFailLoadingWebView(error: WebResourceError) {
        var errorString: String? = error.toString()
        if (errorString == null) {
            errorString = ""
        }
        Log.d(mTagName, "WebView didFailLoadingWebView with Error:\n$errorString")

        if (this.status != VideoLoadStatus.LOAD_FAILED) {
            if (this.playerListener != null) {
                this.playerListener!!.useekPlayerDidFailWithError(this, error)
            }
        }

        this.status = VideoLoadStatus.LOAD_FAILED

        if (!this.isLoadingMaskHidden) {
            this.animateLoadingMaskToHide()
        } else {
            this.mLoadingMaskView.visibility = View.GONE
        }
    }

    /**
     * Animation of showing mLoadingMaskView
     */
    private fun animateLoadingMaskToShow() {
        this.mLoadingMaskView.visibility = View.VISIBLE
    }

    /**
     * Animation of hiding mLoadingMaskView
     */
    private fun animateLoadingMaskToHide() {
        this.mLoadingMaskView.visibility = View.GONE
    }


    enum class VideoLoadStatus {
        NONE, // Not stated loading
        LOAD_STARTED, // Started loading
        LOADED, // Completed loading
        LOAD_FAILED // Failed loading
    }
}
